### ISSUE TITLE
Plugin system for network data tables

### ISSUE THREAD

**Issue by sylvlecl** [powsybl/pypowsybl#102]
* **Do you want to request a *feature* or report a *bug*?**

Feature.

* **What is the current behavior?**

The data series available for each type of equipment are "hard coded" in pypowsybl.

* **What is the expected behavior?**

In order to make available third party IIDM extensions data from python, we could add a plugin system so that extensions provider can define new columns to be appended to equipments dataframes.

Extensions providerq will need to implement a service for that, a minimal spec would be:
```java
interface NetworkExtensionSeriesProvider {
    String getExtensionName();
    DataframeElementType getElementType();
    void addSeries(NetworkDataframeMapperBuilder builder);
}
```

At the time of mappers creation, we will look for all the series provider for the current equipment type through the `ServiceLoader`.

**Note:**
The order of columns should be stable between executions, we could ensure that by sorting providers by their extension name.

* **What is the motivation / use case for changing the behavior?**

Make IIDM extensions data available to python users.

**Comment by sylvlecl** [comment]
Good first candidates to benefit from that feature would be the following extensions:
- `GeneratorActivePowerControl`
- `HvdcAngleDroopActivePowerControl`
- `HvdcOperatorActivePowerRange`

Note that, since those classes are in powsybl-core, we could implement this without any plugin mechanism.
However, it makes sense to treat all extensions equals, and that's a good opportunity to test that mechanism.

**Comment by sylvlecl** [comment]
@CBiasuzzi raised the issue of "what to provide when extensions are absent".

For example, if generator GEN1 has a GeneratorActivePowerControl but GEN2 does not have it, what should be the content of columns for GEN2 ?

For `float` and `string` it's kind of OK: we have `nan` and empty string.
But for `bool` and `int` we don't have that kind of special values. Of course this is a broader issue than just for extensions, as already mentioned in issue #79 .

**An overview of missing values support in numpy and pandas:**
Basically, in both libraries, there is some support of missing values through boolean masks.
- numpy: [masked arrays](https://numpy.org/doc/stable/reference/maskedarray.html)
- pandas: for example [IntegerArray](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.arrays.IntegerArray.html#pandas.arrays.IntegerArray) for integers, which is implemented as a masked array too. This goes together with the special value `pd.NA`. However, all this is mentioned as experimental in the documentation.
- strangely, there is no automatic conversion between numpy masked arrays and pandas masked arrays. When using a numpy masked array to create a series or dataframe, it's converted to float in order to use `nan` for missing values.

In the end, I think that to have clean support for missing values, we should try to use pandas masked arrays, although it's a bit risky since tagged as experimental.

If we decide to implement this, it means that we must support that concept of masked array also on our java and C sides:
- on C side, we could add an optional `bool* mask` to `array` struct
- on java side, we would need to have either additional `exists` methods or use `OptionalInt` or `OptionalBool` when getting values

I would be interested to have your opinion on that possibility @geofjamg .

**Comment by sylvlecl** [comment]
**Another possible approach:**
a completely different approach to solve the issue in the scope of this evolution would be to not provide the extension data in the same table as network elements, but instead in a separate one with only elements which actually HAVE an extension.

This would allow the user to handle those missing values as he wants, if he ever needs to join the tables.

That would mean that we would have a separate python method to get extensions data:
```python
df = network.get_generators_extension('extension-name')
```

**Comment by geofjamg** [comment]
> @CBiasuzzi raised the issue of "what to provide when extensions are absent".
> 
> For example, if generator GEN1 has a GeneratorActivePowerControl but GEN2 does not have it, what should be the content of columns for GEN2 ?
> 
> For `float` and `string` it's kind of OK: we have `nan` and empty string. But for `bool` and `int` we don't have that kind of special values. Of course this is a broader issue than just for extensions, as already mentioned in issue #79 .
> 
> **An overview of missing values support in numpy and pandas:** Basically, in both libraries, there is some support of missing values through boolean masks.
> 
>     * numpy: [masked arrays](https://numpy.org/doc/stable/reference/maskedarray.html)
> 
>     * pandas: for example [IntegerArray](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.arrays.IntegerArray.html#pandas.arrays.IntegerArray) for integers, which is implemented as a masked array too. This goes together with the special value `pd.NA`. However, all this is mentioned as experimental in the documentation.
> 
>     * strangely, there is no automatic conversion between numpy masked arrays and pandas masked arrays. When using a numpy masked array to create a series or dataframe, it's converted to float in order to use `nan` for missing values.
> 
> 
> In the end, I think that to have clean support for missing values, we should try to use pandas masked arrays, although it's a bit risky since tagged as experimental.
> 
> If we decide to implement this, it means that we must support that concept of masked array also on our java and C sides:
> 
>     * on C side, we could add an optional `bool* mask` to `array` struct
> 
>     * on java side, we would need to have either additional `exists` methods or use `OptionalInt` or `OptionalBool` when getting values
> 
> 
> I would be interested to have your opinion on that possibility @geofjamg .

Maybe it would be simpler to not rely on numpy/pandas mask feature but just to manage an applicative mask.
So in addition to the `Series` array, we could just un bitset with a column * row size. 
The memory impact will be neglictable using a bitset and in that case we can put whatever we want in the Series data when value is invalid (like 0 or whatever).
Also the code impact would be quite low because it would be just passing an additional array from java to python.

**Comment by CBiasuzzi** [comment]
In the current approach (branch network_extension_series and the related pull request ) 
 I'm actually adding a dedicated boolean series, for each extension implementation,
 named extensionName_available. When a network element doesn't have the extension
  the element's attribute is set to false, and all the specific extension's attributes are set to some default values.
Python side, a mask on the 'available' column can be used to filter the elements that don't have the extension, e.g. :

```
generators=network.get_generators()
extension_mask=generators.ActivePowerControl_available
print(generators[extension_mask])
```

This approach seems close to what you are suggesting
 bitset based implementation aside. What's your opinion?

**Comment by geofjamg** [comment]
> In the current approach (branch network_extension_series and the related pull request ) I'm actually adding a dedicated boolean series, for each extension implementation, named extensionName_available. When a network element doesn't have the extension the element's attribute is set to false, and all the specific extension's attributes are set to some default values. Python side, a mask on the 'available' column can be used to filter the elements that don't have the extension, e.g. :
> 
> ```
> generators=network.get_generators()
> extension_mask=generators.ActivePowerControl_available
> print(generators[extension_mask])
> ```
> 
> This approach seems close to what you are suggesting bitset based implementation aside. What's your opinion?

I really doubt about this design because, it clutters the data with extra boolean for extension presence. 
Moreover, as pypowsybl is more and more used, we cannot easily go back if we later find a better alternative.
What I think we should do:
 - add an extra mask boolean array (columns asked * element count) to the Java/Python interface to model all absent values.
 - Post process on Python side, to replace absent values by a proper Python None => so mark array is never exposed to users!!!!!

The advantages of this solution:
 - we do not rely on bleeding edge Pandas/Numpy mask feature.
 - we do not change too much the GraalVM C interface (juste adding an extra boolean array)
 - everything is transparent for users

**Comment by sylvlecl** [comment]
>     * Post process on Python side, to replace absent values by a proper Python None => so mark array is never exposed to users!!!!!

What do you mean exactly for that second step ? The issue is that there is no such thing as a proper python None in numpy or pandas arrays (since behind the scene it's plain C arrays). 

What happens when you set a `None` in a dataframe, with for ex:
```python
df.loc['GEN', 'target_p'] = None
```
is that:

- for float arrays, it becomes a `NaN`
- for int arrays, the array is converted to `float` and it becomes a `NaN`
- for boolean arrays, the array is converted to `object`, which indeed contains a `None`. It brings some issues, for instance when you use that array for filtering, since it is not anymore a boolean array.

So it does not seem like a very clean solution. However, it's indeed the "standard" pandas way of handling missing data today, so maybe we should simply use it and let the user work around it when needed, using pandas features.

**Comment by geofjamg** [comment]
@sylvlecl @CBiasuzzi I think I totally change my mind about this design: I am not sure anymore it is a good idea to integrate extension columns to its parents because:
 - it would be hard for users to figure out what are related columns
 - it would add more and more columns to the bases table and for most tables there is already a LOT
 
Maybe it would be cleaner for the plugin system to expose new tables, one per extension with an extra column for the parent ID.
In the network we could dynamically list extensions tables:
```python
n.get_extensions_names()
```

then to get a table
```python
n.get_extension(name)
```

MAIN DRAWBACK users have to known how to join 2 tables, but
 - this is straightforward with Pandas
 -  it could be explained in our doc

What do you all think about this?

**Comment by sylvlecl** [comment]
Indeed it sounds reasonable and simplifies the implementation, although accessing some data will not feel as natural (for ex. generator droop ... ).

It relaxes a little the prerequisite of "missing data handling", since the getters would only return data for existing extensions.

From a general point of view, it's a good way to let the user responsible for choices we don't really want to make: how to handle missing extensions, how to name the columns if there's a conflict, etc.

On the downside, the "default attributes" feature we have implemented to filter out some columns may not make that much sense anymore, if we go this way.

**Comment by sylvlecl** [comment]
Implemented in #277 and improved with modification in #393
